### PR TITLE
notifications v1

### PR DESCRIPTION
**Pull request by AntonPuko**
ezgif-4-fe851ddda9
There are flow errors, and everything is not very clean..but it will be a long road to make it perfect. In general, seems working for all the cases.

Currently, I'm using these requests for notifications:
low keg level:
```
{
  "to":"ewj2Kuk1VTo:APA91bH-7X37EcrK3N5eppq4qpwoIhts1XusWT1sH081opHtZrhi9T0pS5Sn6DoUTs25ODdLMu2aK-_zDxLNmuu2dhLJlG5QnVY8_EFsMcS_gqaeaWVtvunkZefTZhJPdF2R-X2kUV7H",
  "data": {
    "custom_notification": {
      "id": "{{$guid}}",
      "body": "A Beverage level on \"1 - Tubor Lite\" is below 10%!",
      "title": "Low keg level!",
      "color":"#3c91a1",
      "priority":"high",
      "icon":"ic_launcher",
      "large_icon": "ic_launcher",
      "type": "lowKegLevel",
      "tapId": "294",
      "beverageName": "Tuborg Lite",
      "beverageId": "33635"
    }
  }
} 
```
new achievement:
```
{
  "to":"ewj2Kuk1VTo:APA91bH-7X37EcrK3N5eppq4qpwoIhts1XusWT1sH081opHtZrhi9T0pS5Sn6DoUTs25ODdLMu2aK-_zDxLNmuu2dhLJlG5QnVY8_EFsMcS_gqaeaWVtvunkZefTZhJPdF2R-X2kUV7H",
  "data": {
    "custom_notification": {
      "id": "{{$guid}}",
      "body": "Beer Authority",
      "title": "New achievement",
      "color":"#3c91a1",
      "priority":"high",
      "icon":"ic_launcher",
      "large_icon": "ic_launcher",
      "type": "newAchievement",
      "achievementType": "Welcome"
    }
  }
} 
```
new friend request:
```
{
  "to":"ewj2Kuk1VTo:APA91bH-7X37EcrK3N5eppq4qpwoIhts1XusWT1sH081opHtZrhi9T0pS5Sn6DoUTs25ODdLMu2aK-_zDxLNmuu2dhLJlG5QnVY8_EFsMcS_gqaeaWVtvunkZefTZhJPdF2R-X2kUV7H",
  "data": {
    "custom_notification": {
      "id": "{{$guid}}",
      "body": "John wants to be friends with you!",
      "title": "New friend request",
      "color":"#3c91a1",
      "priority":"high",
      "friendUserName": "john",
      "friendId": "68ae3ba1-7841-4273-8d4b-0fb4fa6ed3ca",
      "icon":"ic_launcher",
      "large_icon": "ic_launcher",
      "type": "newFriendRequest"
    }
  }
} 
```

### CODE REVIEW

File: src/stores/NotificationsStore.js [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+export type NewFriendRequestNotification = BaseNotification<
+  NewFriendRequestNotificationExtraProps,
+>;
+
+export type Notification =
+  | LowKegLevelNotification
+  | NewAchievementNotification
+  | NewFriendRequestNotification;
+
+class NotificationsStore {
```
Comment (AntonPuko) [inline src/stores/NotificationsStore.js]: Yeah..I know..this store smells not very well. :shy:

Comment (jlkalberer) [reply src/stores/NotificationsStore.js]: I actually think this is fine. Push notifications are hacky by nature. What you have here seems reasonable.

File: src/stores/NotificationsStore.js [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+      const initialNotification = await FCM.getInitialNotification();
+      initialNotification && this._onRawNotification(initialNotification);
+    }, 500);
+  };
+
+  registerToken = async () => {
+    const fcmToken = await FCM.getFCMToken();
+    const deviceUniqueID = DeviceInfo.getUniqueID();
+
+    await fetchJSON(`${BASE_PUSH_URL}/register`, {
```
Comment (jlkalberer) [inline src/stores/NotificationsStore.js]: I'm going to rethink the API a bit (next week). We'll need a `register` just to connect to FCM but we'll need the ability to unsubscribe from specific entities (taps/friends) and more in the future. This here should stay the same though.

File: src/stores/NotificationsStore.js [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  get hasUnread(): boolean {
+    return this.unreadCount > 0;
+  }
+
+  @computed
+  get _isInitialized(): boolean {
+    return !!this._navigation && AuthStore.isAuthorized;
+  }
+
+  handleInitialNotification = async () => {
```
Comment (jlkalberer) [inline src/stores/NotificationsStore.js]: How do you handle the case where I got 5 notifications without the app open. I open the app without clicking one of those notifications? If it's covered somewhere then I missed it.

File: src/stores/NotificationsStore.js [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    if (openedFromTray) {
+      this.onNotificationPress(parsedNotification);
+    }
+  };
+
+  _handleNotificationPressByType = (notification: Notification) => {
+    switch (notification.type) {
+      case 'lowKegLevel': {
+        const { tapId }: LowKegLevelNotification = (notification: any);
+        KegStore.flushCache();
```
Comment (jlkalberer) [inline src/stores/NotificationsStore.js]: Shouldn't you just flush the keg for a specific tapId? I could even send the kegID in the push notification.

File: src/stores/NotificationsStore.js (hunk 2 above)
Comment (AntonPuko) [reply src/stores/NotificationsStore.js]: yeah. I remember that we need taps opt-out.

File: src/stores/NotificationsStore.js (hunk 3 above)
Comment (AntonPuko) [reply src/stores/NotificationsStore.js]: If 5 notifications came and you open the app, all notifications are on the notifications screen and all removed from the Android tray. If you open the app by clicking on one of those notifications it redirects to what this notif routes to. The other logic is the same as in the 1st case.

File: src/stores/NotificationsStore.js (hunk 4 above)
Comment (AntonPuko) [reply src/stores/NotificationsStore.js]: yeah. it should be more precise, but we don't have flushEntityCacheByID() method in DAO right now.

File: src/stores/NotificationsStore.js (hunk 3 above)
Comment (jlkalberer) [reply src/stores/NotificationsStore.js]: Sweet. That's exactly what we want.

**Pull request merged**